Introduce browser client hint detection

sgiehl opened the pull request:
### Description:

As discussed in some other issues it might happen in the near future that some devices or clients might freeze the useragents in favor of client hints.

This is a proposal on how we could possibly implement a client hint detection in device detector
It's not yet finished and there are still a couple of things that might need to be discussed and improved.

I did not yet had much time to think about this or implement it, so if someone has a suggestion for a better solution feel free to suggest it here.

@sanchezzzhak Any feedback on that one?

sgiehl committed:
Started implementing client hint detection

sgiehl committed:
Started implementing client hint detection

sanchezzzhak reviewed (dismissed):
it looks good

1 we need to added test for windows 11
https://github.com/matomo-org/device-detector/issues/6928#issuecomment-990658109

2 maybe worth adding to readme: if use `client-hints`, you need to send the header yourself (before output html)
```php
header('Accept-CH: Sec-CH-UA-Full-Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform-Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);
```

mimmi20 commented on Parser/AbstractParser.php:
```diff
@@ -42,6 +43,12 @@ abstract class AbstractParser
+     * Holds the user agent the should be parsed
+     * @var ClientHints
+     */
+    protected $clientHits;
```
Shouldn't this be `protected $clientHints;`?

sgiehl replied:
sure. will change that

sgiehl committed:
fix typo

sgiehl committed:
fix typo

sgiehl commented:
@sanchezzzhak If you have some time free, feel free to push some adjustments & tests to this branch. I'm not yet sure if I will find much time this week to go on with this....

sanchezzzhak committed:
feat(clienthints) detect windows 11

sanchezzzhak committed:
feat(clienthints) detect windows 11

sanchezzzhak committed:
fix: phpcs

sanchezzzhak committed:
fix: phpcs

liviuconcioiu commented:
@sanchezzzhak if `$_SERVER['HTTP_X_REQUESTED_WITH']` is going to be added, I will suggest to be added also those:

`$_SERVER['HTTP_DEVICE_STOCK_UA']`
`$_SERVER['HTTP_X_OPERAMINI_PHONE_UA']`
`$_SERVER['HTTP_X_UCBROWSER_UA']`

Examples:

| $_SERVER['HTTP_USER_AGENT'] | $_SERVER['HTTP_DEVICE_STOCK_UA'] | $_SERVER['HTTP_X_OPERAMINI_PHONE_UA'] |
| ------------- | ------------- | ------------- |
| Opera/9.80 (MAUI Runtime; Opera Mini/4.4.39014/191.265; U; hu) Presto/2.12.423 Version/12.16 | OperaMini(Nokia 210FW40.01.11;Opera Mini/4.4.39014;hu) | OperaMini(Nokia 210FW40.01.11;Opera Mini/4.4.39014;hu) |
| Opera/9.80 (MAUI Runtime; Opera Mini/4.4.33576/191.265; U; en) Presto/2.12.423 Version/12.16 | OperaMini(MAUI_MRE;Opera Mini/4.4.33576;en) | OperaMini(MAUI_MRE;Opera Mini/4.4.33576;en) |
| Opera/9.80 (MAUI Runtime; Opera Mini/4.4.33576/191.265; U; en) Presto/2.12.423 Version/12.16 | OperaMini(Fucus/Unknown;Opera Mini/4.4.33576;en) | OperaMini(Fucus/Unknown;Opera Mini/4.4.33576;en) |
| Opera/9.80 (J2ME/MIDP; Opera Mini/4.5.40312/191.265; U; nl) Presto/2.12.423 Version/12.16 | Nokia301.1/2.0 (08.02) Profile/MIDP-2.1 Configuration/CLDC-1.1 | Nokia301.1/2.0 (08.02) Profile/MIDP-2.1 Configuration/CLDC-1.1 |
| Opera/9.80 (J2ME/MIDP; Opera Mini/4.5.40312/191.256; U; es) Presto/2.12.423 Version/12.16 | Nokia302/5.0 (14.78) Profile/MIDP-2.1 Configuration/CLDC-1.1 Mozilla/5.0 AppleWebKit/420+ (KHTML, like Gecko) Safari/420+ | Nokia302/5.0 (14.78) Profile/MIDP-2.1 Configuration/CLDC-1.1 Mozilla/5.0 AppleWebKit/420+ (KHTML, like Gecko) Safari/420+ |
| Opera/9.80 (SpreadTrum; Opera Mini/4.4.33961/191.265; U; en) Presto/2.12.423 Version/12.16 | NativeOperaMini(Spreadtrum/HW Version:        fp6531_bar;Native Opera Mini/4.4.33961;en) | NativeOperaMini(Spreadtrum/HW Version:        fp6531_bar;Native Opera Mini/4.4.33961;en) |
| Opera/9.80 (J2ME/MIDP; Opera Mini/4.1.14287/191.256; U; en) Presto/2.12.423 Version/12.16 | Nokia2330c-2/2.0 (08.20) Profile/MIDP-2.1 Configuration/CLDC-1.1 | Nokia2330c-2/2.0 (08.20) Profile/MIDP-2.1 Configuration/CLDC-1.1 |

| $_SERVER['HTTP_USER_AGENT'] | $_SERVER['HTTP_X_UCBROWSER_UA'] |
| ------------- | ------------- |
| Mozilla/5.0 (Linux; U; Android 10; en-US; VOG-L29 Build/HUAWEIVOG-L29) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/78.0.3904.108 UCBrowser/13.4.0.1306 Mobile Safari/537.36 | dv(VOG-L29);pr(UCBrowser/13.4.0.1306);ov(Android 10);ss(360*715);pi(1080*2147);bt(GJ);pm(1);bv(1);nm(0);im(0);sr(2);nt(1);bi(355);fi(145);dn(32210210557-ff0aa3dd);ai(720089404747345); |

sanchezzzhak commented:
$_SERVER['HTTP_X_UCBROWSER_UA']
it looks interesting, it will be possible to make some hints  of the screen size.
baidu, instagram - also contain something similar in the user agent line

sgiehl commented on Parser/OperatingSystem.php:
```diff
@@ -238,43 +238,73 @@ public static function getShortOsData(string $name): array
-        foreach ($osRegex['versions'] ?? [] as $regex) {
-            $matches = $this->matchUserAgent($regex['regex']);
+        // parse the useragent if os wasn't provided in client hints
+        if (empty($name) || empty($version)) {
```
not sure if we should check for the version here. There might be operating system that don't provide a version in the client hints. Not sure if we should parse the useragent as fallback in that case.
What do you think about parsing the useragent if no version is available in the client hints, but only use the version from useragent if the name matches?

sanchezzzhak replied:
if major version got 0, it means that you don't need client prompts, see UserAgent (this only applies to windows)

Version | platformVersion
-- | --
Win7/8/8.1 | 0
Win10 1507 | 1
Win10 1511 | 2
Win10 1607 | 3
Win10 1703 | 4
Win10 1709 | 5
Win10 1803 | 6
Win10 1809 | 7
Win10 1903 | 8
Win10 1909 | 8
Win10 2004 | 10
Win10 20H2 | 10
Win10 21H1 | 10
Win10 21H2 | 10
Win11 | 13+

undocumented versions
Version | platformVersion
-- | --
win11 beta | 11 
win11 beta rc 1| 12 
win11 beta rc 2| 13

sgiehl commented:
@sanchezzzhak I wouldn't start adding detection for screen sizes or similar for now. Guess there are also client hints for such stuff. Let's focus on bringing the client hints detection for the already detected stuff on the road. We can also consider parsing other header fields later...

sanchezzzhak committed:
fix: windows version

sanchezzzhak committed:
fix: windows version

sanchezzzhak committed:
fix: phpcs

sanchezzzhak committed:
fix: phpcs

sanchezzzhak committed:
fix: tests

sanchezzzhak committed:
fix: tests

sanchezzzhak committed:
feat(test) added tests from opera and edge

sanchezzzhak committed:
feat(test) added tests from opera and edge

sanchezzzhak committed:
fix: phpcs

sanchezzzhak committed:
fix: phpcs

sgiehl committed:
Improve parsing of operating system

sgiehl committed:
Improve parsing of operating system

sgiehl committed:
adds browser parsing for client hints

sgiehl committed:
adds browser parsing for client hints

sgiehl committed:
simplify code

sgiehl committed:
simplify code

sgiehl committed:
more fixes & tests

sgiehl committed:
more fixes & tests

sgiehl commented:
@sanchezzzhak I have added improved some more code. Now also the browser should be detected with client hints if possible. 
Maybe you can have a quick look if everything makes sense so far. If you have some more real test data for client hints, feel free to extend the `clienthints.yml` fixture.

sanchezzzhak committed:
feat(clienthint) create three real test for browser Chrome, Yandex, OperaMobile

sanchezzzhak committed:
feat(clienthint) create three real test for browser Chrome, Yandex, OperaMobile

sanchezzzhak commented on Tests/fixtures/clienthints.yml:
```diff
@@ -0,0 +1,79 @@
+    Sec-CH-UA: '" Not A;Brand";v="99", "Chromium";v="96", "Opera";v="82", "OperaMobile";v="67"'
+    Sec-CH-UA-Platform: "Android"
+    Sec-CH-UA-Mobile: "?1"
+    Sec-CH-UA-Full-Version: "96.0.4664.104"
```
These are 3 tests that I did from my phone.
For some reason, Opera returns the Chrome version, not its own.

sanchezzzhak replied:
I used this site for get information of CH
https://browserleaks.com/client-hints

sgiehl replied:
Tried Opera Mobile in my phone with Android 11. There Opera also returns the chrome version...

sgiehl replied:
Btw. another site to test the headers is https://user-agent-client-hints.glitch.me/

mimmi20 commented on Tests/fixtures/clienthints.yml:
```diff
@@ -0,0 +1,79 @@
+    Sec-CH-UA: '" Not A;Brand";v="99", "Chromium";v="96", "Yandex";v="22"'
+    Sec-CH-UA-Platform: "Android"
+    Sec-CH-UA-Mobile: "?1"
+    Sec-CH-UA-Full-Version: "22.1.4.706"
```
Is there a known reason, why theses versions do not match?

sanchezzzhak replied:
this is not a typo, I double-checked
![image](https://user-images.githubusercontent.com/1337066/155879231-1635b82a-218e-42ac-8ee6-9ffd2dc02525.png)

sanchezzzhak replied:
only today the answer came, from yandex


> Hello!
> Sorry for the delay in responding.
> The versions do not match, because in the User Agent we write the version of the application, and in the client hits a more accurate version of the assembly.

liviuconcioiu commented:
```YML
-
  user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.14335.105", "CCleaner Browser";v="98.0.14335.105"'
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.14335.105"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Model: ""
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Windows
    version: "10"
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.102
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Chrome

```

sgiehl committed:
add more test cases

sgiehl committed:
add more test cases

liviuconcioiu commented:
Test that are correct:
```YAML
-
  user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.14335.104", "AVG Secure Browser";v="98.0.14335.104"'
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.14335.104"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Model: ""
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Windows
    version: "10"
    platform: x86
  client:
    type: browser
    name: AVG Secure Browser
    version: 98.0.14335.104
    engine: ""
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Chrome
-
  user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.14335.103", "Avast Secure Browser";v="98.0.14335.103"'
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.14335.103"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Model: ""
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Windows
    version: "10"
    platform: x86
  client:
    type: browser
    name: Avast Secure Browser
    version: 98.0.14335.103
    engine: ""
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Chrome
```

Here is seen as Chromium, but in the user agent has Silk. Maybe if it has only Chromium and not the browser name in client hints, it should detect from user agent?

```YAML
-
  user_agent: Mozilla/5.0 (Linux; Android 9; KFKAWI) AppleWebKit/537.36 (KHTML, like Gecko) Silk/98.1.148 like Chrome/98.0.4758.101 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.101"'
    Sec-CH-UA-Platform: "Android"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.101"
    Sec-CH-UA-Platform-Version: "9.0.0"
    Sec-CH-UA-Model: "KFKAWI"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Android
    version: 9.0.0
    platform: ""
  client:
    type: browser
    name: Chromium
    version: 98.0.4758.101
    engine: ""
    engine_version: ""
  device:
    type: tablet
    brand: Amazon
    model: Fire HD 8 (2018)
  os_family: Android
  browser_family: Chrome
```

mimmi20 commented on Tests/fixtures/clienthints.yml:
```diff
@@ -77,3 +77,146 @@
+  headers:
+    Sec-CH-UA: "\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"98\", \"Opera\";v=\"84\""
+    Sec-CH-UA-Mobile: "?0"
+    Sec-CH-UA-Full-Version-List: "\" Not A;Brand\";v=\"99.0.0.0\", \"Chromium\";v=\"98.0.4758.82\", \"Opera\";v=\"98.0.4758.82\""
```
Are the Opera versions really this different?

liviuconcioiu replied:
I can also confirm from the data I have, that the versions are different.

sgiehl committed:
more tests

sgiehl committed:
more tests

sgiehl committed:
Adds special handling if cient hints report chromium, but user agent another chromium based browser

sgiehl committed:
Adds special handling if cient hints report chromium, but user agent another chromium based browser

sgiehl commented:
@liviuconcioiu AVG browser doesn't have an engine defined:
https://github.com/matomo-org/device-detector/blob/d61e6e93b7f8ec6ed8fb895f1bc1a2c2bfac18a2/regexes/client/browsers.yml#L625-L629
same for Avast:
https://github.com/matomo-org/device-detector/blob/d61e6e93b7f8ec6ed8fb895f1bc1a2c2bfac18a2/regexes/client/browsers.yml#L987-L991

If both are Chromium based, feel free to add an engine there in another PR.

I have also added a special handling for chromium in case the user agent reports another chromium based browser.

liviuconcioiu commented:
Platform should be only ARM and for mobile version the browser name is put in front before Chromium and is not detected.

```YAML
-
  user_agent: Mozilla/5.0 (Linux; Android 11; sohKwhigbQ; U; en) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.101 Mobile AvastSecureBrowser/6.6.0 Build/3850 Safari/537.36
  headers:
    Sec-CH-UA: '"AvastSecureBrowser";v="6.6.0", " Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.101"'
    Sec-CH-UA-Platform: "Android"
    Sec-CH-UA-Mobile: "?1"
    Sec-CH-UA-Full-Version: "6.6.0"
    Sec-CH-UA-Platform-Version: "11"
    Sec-CH-UA-Arch: "[arm64-v8a, armeabi-v7a, armeabi]"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Android
    version: "11"
    platform: "[arm64-v8a, armeabi-v7a, armeabi]"
  client:
    type: browser
    name: Chromium
    version: 6.6.0
    engine: ""
    engine_version: ""
  device:
    type: tablet
    brand: ""
    model: ""
  os_family: Android
  browser_family: Chrome
```

sgiehl commented:
Hm. I've just realized that the order of the brands in the ua list actually can vary. Further more the specs suggest to let them vary over time to protect fingerprints using them. So guess I need to adjust the detection a bit more...

sgiehl committed:
further tests & improvements

sgiehl committed:
further tests & improvements

sgiehl commented:
@liviuconcioiu won't chrome itself also possibly hide the `Google Chrome` (sometimes)? Currently it detects the browser from useragent if the the client hints are reporting `Chromium`. If also pushed several other improvements. Let me know if you have any other suggestions.

liviuconcioiu commented on Tests/fixtures/clienthints.yml:
```diff
@@ -298,3 +298,55 @@
+    model: ""
+  os_family: Android
+  browser_family: Chrome
+-
+  user_agent: ""
+  headers:
+    Sec-CH-UA: '"AvastSecureBrowser";v="6.6.0", " Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.101"'
+    Sec-CH-UA-Platform: "Android"
+    Sec-CH-UA-Mobile: "?1"
+    Sec-CH-UA-Full-Version: "6.6.0"
+    Sec-CH-UA-Platform-Version: "11"
+    Sec-CH-UA-Arch: "[arm64-v8a, armeabi-v7a, armeabi]"
+    Sec-CH-Prefers-Color-Scheme: "light"
+  os:
+    name: Android
+    version: "11"
+    platform: "ARM"
+  client:
+    type: browser
+    name: Avast Secure Browser
+    version: "6.6.0"
+    engine: ""
+    engine_version: ""
+  device:
+    type: ""
+    brand: ""
+    model: ""
+  os_family: Android
+  browser_family: Chrome
```
This test is a duplicate of the one above.

sgiehl replied:
No. It lacks the useragent. Added that to ensure not everything comes from the useragent fallback, which was the case due to some incorrect code before...

liviuconcioiu commented:
Missing platform:

```YAML
-
  user_agent: Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36 Edg/98.0.1108.56
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.1108.56", "Microsoft Edge";v="98.0.1108.56"'
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.1108.56"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Bitness: "32"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Windows
    version: "10"
    platform: ""
  client:
    type: browser
    name: Microsoft Edge
    version: 98.0.1108.56
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Internet Explorer
-
  user_agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.109 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.109", "Google Chrome";v="98.0.4758.109"'
    Sec-CH-UA-Platform: "macOS"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.109"
    Sec-CH-UA-Platform-Version: "10.14.6"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: "dark"
  os:
    name: Mac
    version: 10.14.6
    platform: ""
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.109
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: Apple
    model: ""
  os_family: Mac
  browser_family: Chrome
```

Missing platform and macOS versions don't match (I think the OS version is frozen in user agent at 10.15.7 - from what I see in other user agents/client hints, so it should get the greater version if don't match?):

```YAML
-
  user_agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36 OPR/84.0.4316.21
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.102", "Opera";v="98.0.4758.102"'
    Sec-CH-UA-Platform: "macOS"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.102"
    Sec-CH-UA-Platform-Version: "11.6.0"
    Sec-CH-UA-Arch: "arm"
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Mac
    version: 10.15.7
    platform: ""
  client:
    type: browser
    name: Opera
    version: 98.0.4758.102
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: Apple
    model: ""
  os_family: Mac
  browser_family: Opera
```

Missing Linux version:

```YAML
-
  user_agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.102"'
    Sec-CH-UA-Platform: "Linux"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.102"
    Sec-CH-UA-Platform-Version: "5.4.0"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: GNU/Linux
    version: ""
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.102
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: GNU/Linux
  browser_family: Chrome
```

For devices that aren't added, there is no model name:

```YAML
-
  user_agent: Mozilla/5.0 (Linux; Android 10; X30) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.101 Mobile Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.101", "Google Chrome";v="98.0.4758.101"'
    Sec-CH-UA-Platform: "Android"
    Sec-CH-UA-Mobile: "?1"
    Sec-CH-UA-Full-Version: "98.0.4758.101"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Model: "X30"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Android
    version: 10.0.0
    platform: ""
  client:
    type: browser
    name: Chrome Mobile
    version: 98.0.4758.101
    engine: Blink
    engine_version: ""
  device:
    type: smartphone
    brand: ""
    model: ""
  os_family: Android
  browser_family: Chrome
```

sgiehl committed:
further tests & improvements

sgiehl commented:
@liviuconcioiu Thanks for that. I have added all the tests and fixed the code so the results are correct. Maybe you can have a quick look over them if everything is correct now.

sanchezzzhak commented on Parser/OperatingSystem.php:
```diff
@@ -361,13 +344,146 @@ public static function getNameFromId(string $os, ?string $ver = null): ?string
+        if ($this->clientHints instanceof ClientHints && $this->clientHints->getOperatingSystem()) {
+            $hintName = $this->clientHints->getOperatingSystem();
+
+            if ('macOS' === $hintName) {
+                $hintName = 'Mac';
+            } elseif ('Linux' === $hintName) {
+                $hintName = 'GNU/Linux';
+            }
```
maybe we should make an array?
```php
$hintOsNames = ['macOS' => 'Mac', 'Linux' => 'GNU/Linux' ];
$hintName = $hintOsNames[$hintName] ?? $hintName;
```
can also make all keys in lowercase

sgiehl replied:
Was also thinking of some kind of static mapping array. Same for browsers... Guess over time we will need to add more entries there, as soon as more browsers start supporting it

liviuconcioiu commented:
@sgiehl Everything seems to be ok so far.

sgiehl committed:
use static mapping arrays

liviuconcioiu commented:
If Chrome OS doesn't report version in client hints, then the os version is taken from Chrome.
```YAML
-
  user_agent: Mozilla/5.0 (X11; CrOS x86_64 14388.61.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.107 Safari/537.36
  headers:
    Sec-CH-UA-Platform: "Chrome OS"
    Sec-CH-UA-Mobile: "?0"
  os:
    name: Chrome OS
    version: 98.0.4758.107
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.107
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Chrome OS
  browser_family: Chrome
```

sgiehl commented:
> If Chrome OS doesn't report version in client hints, then the os version is taken from Chrome.

That happens because it uses the version detected from user agent in that case. What would you expect here?

liviuconcioiu commented:
> > If Chrome OS doesn't report version in client hints, then the os version is taken from Chrome.
> 
> That happens because it uses the version detected from user agent in that case. What would you expect here?

`14388.61.0`

sanchezzzhak committed:
Merge remote-tracking branch 'upstream/clienthints' into clienthints

# Conflicts:
#	ClientHints.php
#	Parser/Client/Browser.php
#	Parser/OperatingSystem.php
#	Tests/fixtures/clienthints.yml

sgiehl commented:
Should we handle `Opera Crypto` as new browser, or maybe simply map it to `Opera` instead?

Regarding Xbox. That is kind of on purpose. There is currently not further handling for the model names. If the useragent doesn't provide any details on the model / brand then the model from client hints will be used (with an empty brand).
Not sure if we should try to perform further checks in that case. We could e.g. try to create a "fake" useragent based on the client hints and try to parse that one in order to get device and brand. But not sure if that won't result in incorrect detection.

liviuconcioiu commented:
> @liviuconcioiu so `Opera Crypto` can't be detected from user agent only?
> 
> The model for xbox should correctly be set. And it's detected as desktop, because no model / device can be detected from useragent and the operating system is a desktop only one...

No, Opera Crypto cannot be detect from user agent.

sgiehl commented:
Ok. What about having an additional class constant like `clientHintsOnlyBrowsers`, which holds all browsers names that are coming from client hints only. We could then ignore those browsers in the tests

sgiehl committed:
Adds detection for Opera Crypto

sanchezzzhak committed:
Merge remote-tracking branch 'upstream/clienthints' into clienthints

sanchezzzhak committed:
fix: phpcs

liviuconcioiu commented:
`Gener8` browser - [https://gener8ads.com/products/browser](https://gener8ads.com/products/browser)

```YAML
-
  user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.109 Safari/537.36
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99.0.0.0", "Chromium";v="98.0.4758.109", "Gener8";v="98.0.4758.109"'
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.109"
    Sec-CH-UA-Platform-Version: "10.0.0"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Windows
    version: "10"
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.109
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Chrome
```

sgiehl committed:
Adds detection for Gener8 browser

sgiehl committed:
Use detected engine from useragent if browser name is different but family matches

liviuconcioiu commented:
It seems there are even bots that use client hints. If useful for tests:

```YAML
-
  user_agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36 PTST/220216.200657
  headers:
    Sec-CH-UA-Platform: "Unknown"
    Sec-CH-UA-Mobile: "?0"
  bot:
    name: WebPageTest
    category: Site Monitor
    url: https://www.webpagetest.org
-
  user_agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36 PTST/220216.200657
  headers:
    Sec-CH-UA: '" Not A;Brand";v="99", "Chromium";v="98.0.4758.102", "Google Chrome";v="98.0.4758.102"'
    Sec-CH-UA-Platform: "Unknown"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "98.0.4758.102"
    Sec-CH-UA-Platform-Version: "0"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Model: "Model"
    Sec-CH-Prefers-Color-Scheme: "light"
  bot:
    name: WebPageTest
    category: Site Monitor
    url: https://www.webpagetest.org
```

Right now, we detect only `Chrome OS` which is built by Google. `Chromium OS` which anyone can fork and build, can't be detected from user agent only, so it should be done like `Opera Crypto`.

```YAML
-
  user_agent: Mozilla/5.0 (X11; CrOS x86_64 14150.90.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.126 Safari/537.36
  headers:
    Sec-CH-UA-Platform: "Chromium OS"
    Sec-CH-UA-Mobile: "?0"
  os:
    name: Chrome OS
    version: 94.0.4606.126
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 94.0.4606.126
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Chrome OS
  browser_family: Chrome
-
  user_agent: Mozilla/5.0 (X11; CrOS x86_64 14150.90.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.126 Safari/537.36
  headers:
    Sec-CH-UA-Platform: "Chromium OS"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "94.0.4606.126"
    Sec-CH-UA-Platform-Version: "14150.90.0"
    Sec-CH-UA-Arch: "x86"
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: "light"
  os:
    name: Chrome OS
    version: 94.0.4606.126
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 94.0.4606.126
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Chrome OS
  browser_family: Chrome
```

liviuconcioiu commented:
Platform is Linux, but from version and user agent the os name should be Chromium OS instead of GNU/Linux.

```YAML
-
  user_agent: Mozilla/5.0 (X11; CrOS armv7l 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.98 Safari/537.36
  headers:
    Sec-CH-UA-Platform: Linux
    Sec-CH-UA-Mobile: ?0
    Sec-CH-UA-Full-Version: 92.0.4515.98
    Sec-CH-UA-Platform-Version: 13597.84.0
    Sec-CH-UA-Arch: arm
  os:
    name: GNU/Linux
    version: 13597.84.0
    platform: ARM
  client:
    type: browser
    name: Chrome
    version: 92.0.4515.98
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: GNU/Linux
  browser_family: Chrome
```

sgiehl committed:
adds detection for chromium os & various code improvements

sgiehl committed:
fix chrome os reported as linux

sgiehl committed:
Merge remote-tracking branch 'origin/master' into clienthints

sgiehl committed:
fix test

sgiehl commented:
@liviuconcioiu adjusted the code so the above examples are detected correctly.

@liviuconcioiu @sanchezzzhak Would you mind doing a review of the code? If everything looks fine so far I would be tempted to merge this soon. When more stuff needs to be adjusted, we can always do that later

sanchezzzhak commented on Parser/Client/Browser.php:
```diff
@@ -506,6 +518,138 @@ public static function isMobileOnlyBrowser(string $browser): bool
+            $brands = $this->clientHints->getBrandList();
+
+            foreach ($brands as $brand => $brandVersion) {
+                foreach (self::$clientHintMapping as $browserName => $clientHints) {
+                    foreach ($clientHints as $clientHint) {
+                        if (\strtolower($brand) === \strtolower($clientHint)) {
+                            $brand = $browserName;
+
+                            break 2;
+                        }
+                    }
+                }
```
I would put this code in a separate method

```php
 move to AbstractClass
 create static var AbstractClass::$clientHintMapping
 protected function findNameByHintMapping(string $name): string 
    {
        foreach (self::$clientHintMapping as $mapName => $clientHints) {
            foreach ($clientHints as $clientHint) {
                if (\strtolower($name) === \strtolower($clientHint)) {
                    return $mapName;
                }
            }
        }
        return $name;
    }
}
```

```php
foreach ($brands as $brand => $brandVersion) {
    $brand = $this->findNameByHintMapping($brand);
}
```

sanchezzzhak replied:
the same code can be used in OS Parser

sgiehl replied:
will change that.

liviuconcioiu commented on Tests/fixtures/clienthints.yml:
```diff
@@ -0,0 +1,662 @@
+    name: WebPageTest
+    category: Site Monitor
+    url: https://www.webpagetest.org
+-
+  user_agent: Mozilla/5.0 (X11; CrOS x86_64 14150.90.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.126 Safari/537.36
+  headers:
+    Sec-CH-UA-Platform: "Chromium OS"
+    Sec-CH-UA-Mobile: "?0"
+  os:
+    name: Chromium OS
+    version:
+    platform: x64
```
Here, version is not detected.

liviuconcioiu replied:
And, regarding Opera, it should get the version from user agent, rather than client hints, since that is the real version.

sgiehl replied:
Ok. changed the code, so the version from useragent is used if none is provided, but the os family matches.
We need to see if that might cause any incorrect detections in the future...

sgiehl committed:
move client hint mapping to abstract class

sgiehl committed:
use os version from useragent if os family matches

liviuconcioiu commented:
Another test that should be added (a fake Mac user agent on Windows).

```YAML
-
  user_agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.6167.5881 Safari/537.36
  headers:
    Sec-CH-UA-Platform: "Windows"
    Sec-CH-UA-Mobile: "?0"
    Sec-CH-UA-Full-Version: "?1"
  os:
    name: Windows
    version: ""
    platform: ""
  client:
    type: browser
    name: Chrome
    version: 96.0.6167.5881
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: Windows
  browser_family: Chrome
```

Here, it should detect `Fedora` as os.

```YAML
-
  user_agent: Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.121 Safari/537.36
  headers:
    Sec-CH-UA: "Chromium";v="98.0.4758.121", " Not A;Brand";v="99.0.0.0"
    Sec-CH-UA-Platform: Linux
    Sec-CH-UA-Mobile: ?0
    Sec-CH-UA-Full-Version: 98.0.4758.121
    Sec-CH-UA-Platform-Version: 5.16.11
    Sec-CH-UA-Arch: x86
    Sec-CH-UA-Bitness: "64"
    Sec-CH-Prefers-Color-Scheme: light
  os:
    name: GNU/Linux
    version: 5.16.11
    platform: x64
  client:
    type: browser
    name: Chrome
    version: 98.0.4758.121
    engine: Blink
    engine_version: ""
  device:
    type: desktop
    brand: ""
    model: ""
  os_family: GNU/Linux
  browser_family: Chrome
```

sanchezzzhak committed:
chore: Merge remote-tracking branch 'upstream/master' into clienthints

sgiehl committed:
improve os detection

sgiehl commented:
@liviuconcioiu added both.

sgiehl committed:
Merge remote-tracking branch 'origin/master' into clienthints

sgiehl commented:
@liviuconcioiu @sanchezzzhak I'll now release a last version of device detector without client hint support. And afterwards merge this PR. The following release will then be a new major release with client hint support

sgiehl committed:
Merge branch 'master' into clienthints

Nevercold commented:
Hey, I just tried it out.
However, when I enable ClientHints, I still get Windows 10 out instead of 11. Is that supposed to be?

sv3t0sl4v commented:
> Hey, I just tried it out. However, when I enable ClientHints, I still get Windows 10 out instead of 11. Is that supposed to be?

Could you provide your Windows 11 user agent, so I can test?

sgiehl commented:
@Nevercold did you try that on https://devicedetector.lw1.at/ ?

sanchezzzhak commented:
in order for client prompts to work, you need to manually send headers
```php
header('Accept-CH: Sec-CH-UA-Full Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);
```

Nevercold commented:
> @Nevercold did you try that on https://devicedetector.lw1.at/ ?
![image](https://user-images.githubusercontent.com/32977558/157641238-e3057b30-a312-4141-b0ec-53189310c28c.png)




> client hints to work, you need to manually send headers
> 
> ```
> header('Accept-CH: Sec-CH-UA-Full Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);
> ```
I have set.



> > Hey, I just tried it out. However, when I enable ClientHints, I still get Windows 10 out instead of 11. Is that supposed to be?
> 
> Could you provide your Windows 11 user agent, so I can test?
```Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36```

Nevercold commented:
Raw from DeviceDetector:

```{
  "isBot": false,
  "clientInfo": {
    "type": "browser",
    "name": "Chrome",
    "short_name": "CH",
    "version": "99.0.4844.51",
    "engine": "Blink",
    "engine_version": "",
    "family": "Chrome"
  },
  "browserFamily": "Chrome",
  "isMobileOnlyBrowser": false,
  "osInfo": {
    "name": "Windows",
    "short_name": "WIN",
    "version": "10",
    "platform": "x64",
    "family": "Windows"
  },
  "osFamily": "Windows",
  "device": 0,
  "deviceName": "desktop",
  "deviceBrand": {
    "name": "",
    "short_name": ""
  },
  "model": "",
  "icons": {
    "browser": "/icons/browsers/CH.png",
    "os": "/icons/os/WIN.png",
    "device": "/icons/devices/desktop.png",
    "brand": null
  },
  "clientHints": {
    "architecture": "",
    "bitness": "",
    "mobile": false,
    "model": "",
    "platform": "",
    "platformVersion": "",
    "uaFullVersion": "",
    "fullVersionList": {}
  },
  "headers": null
}```

liviuconcioiu commented:
@Nevercold You need to check Consider client hints then click detect.

Nevercold commented:
> @Nevercold You need to check Consider client hints then click detect.
![image](https://user-images.githubusercontent.com/32977558/157642026-5c01b0b9-2ca2-4931-b70d-c33c371b0155.png)

Okay, then Windows 11 comes out, but it still doesn't work for me.

liviuconcioiu commented:
You must send the accept headers as @sanchezzzhak said. Otherwise won't work.

Nevercold commented:
> You must send the accept headers as @sanchezzzhak said. Otherwise won't work.
Have already.

liviuconcioiu commented:
> > You must send the accept headers as @sanchezzzhak said. Otherwise won't work.
> 
> Have already.

You need to add $dd->parse();

```
$clientHints = DeviceDetector\ClientHints::factory($_SERVER);
$dd = new DeviceDetector($userAgent, $clientHints);
$dd->parse();
print_r($dd->getOs());
```

Nevercold commented:
> > > You must send the accept headers as @sanchezzzhak said. Otherwise won't work.
> > 
> > 
> > Have already.
> 
> You need to add $dd->parse();
> 
> ```
> $clientHints = DeviceDetector\ClientHints::factory($_SERVER);
> $dd = new DeviceDetector($userAgent, $clientHints);
> $dd->parse();
> print_r($dd->getOs());
> ```

Was already there, I just forgot to include it in the code, sorry

Nevercold commented:
```      
$clientHints = ClientHints::factory($_SERVER);
var_dump($clientHints->getOperatingSystemVersion());
var_dump($clientHints->getOperatingSystem());
```
Outputs Windows and 14.0.0, so it should work.

liviuconcioiu commented:
> ```
> $clientHints = ClientHints::factory($_SERVER);
> var_dump($clientHints->getOperatingSystemVersion());
> var_dump($clientHints->getOperatingSystem());
> ```
> 
> Outputs Windows and 14.0.0, so it should work.

To get Windows 11, you need to get info from the parsed user agent and client hints.

```
$osInfo = $dd->getOs();
echo $osInfo['version'];
```

Nevercold commented:
but, $osInfo gives me
```
Array
--
  | (
  | [name] => Windows
  | [short_name] => WIN
  | [version] => 10
  | [platform] => x64
  | [family] => Windows
  | )
  ```

Nevercold commented:
Full code:
```php
  public function getDevice (string $userAgent): array
  {
      $clientHints = ClientHints::factory($_SERVER);
      var_dump($clientHints->getOperatingSystemVersion());
      var_dump($clientHints->getOperatingSystem());
      
      $dd = new DeviceDetector($userAgent, $clientHints);
 
      $dd->skipBotDetection();
      $dd->parse();
      
      $osInfo = $dd->getOs();
      print_r($osInfo);
  }
      ```
      
      index: ``` 
    header('Accept-CH: Sec-CH-UA-Full-Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform-Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);
    ```

liviuconcioiu commented:
Can you check if those keys exist and what are the values?

```
$_SERVER['HTTP_SEC_CH_UA_FULL_VERSION']
$_SERVER['HTTP_SEC_CH_UA_PLATFORM']
$_SERVER['HTTP_SEC_CH_UA_PLATFORM_VERSION']
$_SERVER['HTTP_SEC_CH_UA_ARCH']
$_SERVER['HTTP_SEC_CH_UA_MODEL']
```

Nevercold commented:
> $_SERVER['HTTP_SEC_CH_UA_FULL_VERSION']
> $_SERVER['HTTP_SEC_CH_UA_PLATFORM']
> $_SERVER['HTTP_SEC_CH_UA_PLATFORM_VERSION']
> $_SERVER['HTTP_SEC_CH_UA_ARCH']
> $_SERVER['HTTP_SEC_CH_UA_MODEL']

![image](https://user-images.githubusercontent.com/32977558/157650746-5fd52e59-2fb3-434e-8e81-c5ca2a230ced.png)

Nevercold commented:
maybe it's because of how I included the version?
Have this via Composer with the current git-commit.

"matomo/device-detector": "dev-master#a7d5eb349c9f930944fed7ed1a7abea604ea382e",

liviuconcioiu commented:
Can you try the code exactly as it is here https://github.com/matomo-org/device-detector#usage and add`header('Accept-CH: Sec-CH-UA-Full-Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform-Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);` at the top?

It should work.

Nevercold commented:
```php
<?php

use DeviceDetector\ClientHints;
use DeviceDetector\DeviceDetector;
use DeviceDetector\Parser\Device\AbstractDeviceParser;

require __DIR__ . "/../../vendor/autoload.php";
header('Accept-CH: Sec-CH-UA-Full-Version, Sec-CH-UA-Platform, Sec-CH-UA-Platform-Version, Sec-CH-UA-Model, Sec-CH-UA-Arc', true);

#AbstractDeviceParser::setVersionTruncation(AbstractDeviceParser::VERSION_TRUNCATION_NONE);

$userAgent = $_SERVER['HTTP_USER_AGENT'];
$clientHints = ClientHints::factory($_SERVER);
$dd = new DeviceDetector($userAgent, $clientHints);

$dd->parse();
$osInfo = $dd->getOs();
print_r($osInfo);


die();
```

Output: Array ( [name] => Windows [short_name] => WIN [version] => 10 [platform] => x64 [family] => Windows )
But why D:?

Findus23 commented:
I'm pretty sure that's the same issue I also had when updating devicedetector.lw1.at:

`DeviceDetector\ClientHints::factory` looks at the $_SERVER variables, but they are named differently on my setup (nginx) than what the factory expects. Therefore it doesn't detect any of the client hints (you can also see it above in
```
  "clientHints": {
    "architecture": "",
    "bitness": "",
    "mobile": false,
    "model": "",
    "platform": "",
    "platformVersion": "",
    "uaFullVersion": "",
    "fullVersionList": {}
  },
```)

Nevercold commented:
Interesting.
So as web server I use Apache2 with PHP8.1-FPM
Maybe this is the reason.

liviuconcioiu commented:
Sometimes, the platform version is not sent on the first request, it needs a refresh. At least this is how Chrome behaves for example.
![chrome](https://user-images.githubusercontent.com/25424343/157657922-8ffd559a-0645-4e1b-9e64-e2c6ee1e7c41.png)

Nevercold commented:
> I assume that's FPM-related. I'll look into fixing that.

i have disabled fpm, but it is still the same.

Nevercold commented:
and the headers are set, so very weird.

Findus23 commented:
It turns out @sgiehl was faster and already fixed the issue yesterdays:
https://github.com/matomo-org/device-detector/commit/eae679185d2c3dc4c88224e936d4db01b54c4529

So make sure you are on the latest master branch.

And when in doubt do a `var_dump($clientHints);` and compare it with a `var_dump($_SERVER);` to see if the headers are correctly detected.

sanchezzzhak commented:
example with the last commit  dev-master
https://phpsandbox.io/e/x/mfnke?layout=EditorPreview&defaultPath=%2F&theme=dark&showExplorer=no&openedFiles=
I don't have windows11 to check

Nevercold commented:
> It turns out @sgiehl was faster and already fixed the issue yesterdays: [eae6791](https://github.com/matomo-org/device-detector/commit/eae679185d2c3dc4c88224e936d4db01b54c4529)
> 
> So make sure you are on the latest master branch.
> 
> And when in doubt do a `var_dump($clientHints);` and compare it with a `var_dump($_SERVER);` to see if the headers are correctly detected.

It is recognized correctly, but still only gives out windows 10. 
the current commit is pulled (the latest specified and saved in Composer.json )

```php

object(DeviceDetector\ClientHints)#3 (8) {
--
  | ["architecture":protected]=>
  | string(0) ""
  | ["bitness":protected]=>
  | string(0) ""
  | ["mobile":protected]=>
  | bool(false)
  | ["model":protected]=>
  | string(2) """"
  | ["platform":protected]=>
  | string(9) ""Windows""
  | ["platformVersion":protected]=>
  | string(8) ""14.0.0""
  | ["uaFullVersion":protected]=>
  | string(14) ""99.0.4844.51""
  | ["fullVersionList":protected]=>
  | array(3) {
  | [0]=>
  | array(2) {
  | ["brand"]=>
  | string(12) " Not A;Brand"
  | ["version"]=>
  | string(2) "99"
  | }
  | [1]=>
  | array(2) {
  | ["brand"]=>
  | string(8) "Chromium"
  | ["version"]=>
  | string(2) "99"
  | }
  | [2]=>
  | array(2) {
  | ["brand"]=>
  | string(13) "Google Chrome"
  | ["version"]=>
  | string(2) "99"
  | }
  | }
  | }
```

Nevercold commented:
> example with the last commit dev-master https://phpsandbox.io/e/x/mfnke?layout=EditorPreview&defaultPath=%2F&theme=dark&showExplorer=no&openedFiles= I don't have windows11 to check

same.
![image](https://user-images.githubusercontent.com/32977558/157660779-389cf5f6-91a3-418e-b7d1-137f9f9a80db.png)

liviuconcioiu commented:
@Nevercold Can you refresh phpsandbox page?

Nevercold commented:
> @Nevercold Can you refresh phpsandbox page? Chrome sends the platform version on a second request. So this is the issue.

I already have a few times.
Its the same.

liviuconcioiu commented:
On first request:

![first request](https://user-images.githubusercontent.com/25424343/157661463-a85dd7a0-22c0-4e3c-80e5-d0b6e28e4c71.png)

On second request (page refresh)

![second request](https://user-images.githubusercontent.com/25424343/157661476-30a4c982-1817-49a5-9bb5-6f5f0d6b2d31.png)

Nevercold commented:
But also funny that it first says x64 and then x86.

liviuconcioiu commented:
> But also funny that it first says x64 and then x86.

@sgiehl Can you check this when you have time? It should get the platform from bitness.

Nevercold commented:
and in general, it does not work properly either. xD

liviuconcioiu commented:
> and in general, it does not work properly either. xD

This is a browser issue. Headers aren't sent by the browser on the first request. Probably to avoid fingerprinting or cloaking on the first attempt.

sv3t0sl4v commented:
> > and in general, it does not work properly either. xD
> 
> This is a browser issue. Headers aren't sent by the browser on the first request. Probably to avoid fingerprinting or cloaking on the first attempt.

Tricky! Private browsing or in general?

Nevercold commented:
> > and in general, it does not work properly either. xD
> 
> This is a browser issue. Headers aren't sent by the browser on the first request. Probably to avoid fingerprinting or cloaking on the first attempt.

But I have reloaded it many times, so it should be there then.
The version he does get. (14.0.0.)

sanchezzzhak commented:
additional quotes are present, most likely, this does not allow  determine the version

need fix

Nevercold commented:
true, it takes "14.0.0" as version.
outputs at least $clientHints.

Nevercold commented:
Fixed. https://github.com/matomo-org/device-detector/pull/7050

Works now for me.

sgiehl commented:
Client Hints are actually designed in a way that by default only some basic data (Sec-CH-UA, Sec-CH-UA-Mobile, Sec-CH-UA-Platform) is provided in the request headers. If the server requests more details by sending a header to the client the next request will contain all information. So the first request can never detect windows 11 actually, as the platform version isn't available.

Nevercold commented:
> Client Hints are actually designed in a way that by default only some basic data (Sec-CH-UA, Sec-CH-UA-Mobile, Sec-CH-UA-Platform) is provided in the request headers. If the server requests more details by sending a header to the client the next request will contain all information. So the first request can never detect windows 11 actually, as the platform version isn't available.

Thanks for the explanation  :)

nunoperalta commented:
Hey - Thanks very much for this.
Just realized this isn't included in 5.0.5

Any estimate date of when will the next version be released? :)
Thank you very much.

sgiehl commented:
We are currently preparing the implementation in Matomo to see if any further changes are required here. Once that is done we will release a new major version of device detector. Guess within one or two weeks
